velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -153,6 +155,123 @@ TpchPlan TpchQueryBuilder::getQ1Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ3Plan() const {
+  static const std::string kCustomer = "customer";
+  static const std::string kOrders = "orders";
+  static const std::string kLineitem = "lineitem";
+
+  std::vector<std::string> lineitemColumns = {
+      "l_shipdate", "l_orderkey", "l_extendedprice", "l_discount"};
+  std::vector<std::string> ordersColumns = {
+      "o_orderdate", "o_shippriority", "o_custkey", "o_orderkey"};
+  std::vector<std::string> customerColumns = {"c_custkey", "c_mktsegment"};
+
+  auto lineitemSelectedRowType = getRowType(kLineitem, lineitemColumns);
+  const auto& lineitemFileColumns = getFileColumnNames(kLineitem);
+  auto ordersSelectedRowType = getRowType(kOrders, ordersColumns);
+  const auto& ordersFileColumns = getFileColumnNames(kOrders);
+  auto customerSelectedRowType = getRowType(kCustomer, customerColumns);
+  const auto& customerFileColumns = getFileColumnNames(kCustomer);
+
+  // dateVal :2 = '1995-03-15'
+  const auto orderDate = "o_orderdate";
+  const auto shipDate = "l_shipdate";
+  std::string orderDateFilter;
+  std::string shipDateFilter;
+  std::string customerFilter;
+
+  // DWRF does not support Date type. Use Varchar instead.
+  if (ordersSelectedRowType->findChild(orderDate)->isVarchar()) {
+    orderDateFilter = "o_orderdate < '1995-03-15'";
+  } else {
+    orderDateFilter = "o_orderdate < '1995-03-15'::DATE";
+  }
+
+  if (lineitemSelectedRowType->findChild(shipDate)->isVarchar()) {
+    shipDateFilter = "l_shipdate > '1995-03-15'";
+  } else {
+    shipDateFilter = "l_shipdate > '1995-03-15'::DATE";
+  }
+
+  // MarketSegment filter value :1 = 'BUILDING'
+  customerFilter = "c_mktsegment = 'BUILDING'";
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId lineitemPlanNodeId;
+  core::PlanNodeId ordersPlanNodeId;
+  core::PlanNodeId customerPlanNodeId;
+
+  auto customers = PlanBuilder(planNodeIdGenerator)
+                       .tableScan(
+                           kCustomer,
+                           customerSelectedRowType,
+                           customerFileColumns,
+                           {customerFilter})
+                       .capturePlanNodeId(customerPlanNodeId)
+                       .planNode();
+
+  auto orders =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(
+              kOrders,
+              ordersSelectedRowType,
+              ordersFileColumns,
+              {orderDateFilter})
+          .capturePlanNodeId(ordersPlanNodeId)
+          .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .localPartition(
+              {},
+              {PlanBuilder(planNodeIdGenerator)
+                   .tableScan(
```

Comment:
I wonder if this is the optimal join order. Looking at sf1, I see the following:

- # customer rows with mktsegment = 'BUILDING' - 337
- # orders with orderdate < date '1995-03-15' - 7286
- # lineitems with shipdate > date '1995-03-15' - 32260

Hence, it would make sense to join orders with customer first (with customers on the build side), then join with lineitem (with lineitem on the probe side). The first join will return 1797 rows, which is a lot less than number of lineitem rows.

Reply:
Thanks for the suggestion. The join order has been changed accordingly and with 4 drivers, ~21% savings are seen compared to the time measured earlier. I will share the detailed profiling data shortly (the build is failing currently because of duckDb parquet reader movement changes).